### ISSUE TITLE
First-class CORS support/helpers

### ISSUE THREAD

**Issue by kgriffs** [falconry/falcon#303]
Should this be done using hooks? See #238 and #246.

Note: Include preflight support...

Reference: 
* https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS?redirectlocale=en-US&redirectslug=HTTP_access_control
* http://www.html5rocks.com/en/tutorials/cors/

**Comment by lwcolton** [comment]
I'll work on this

**Comment by lwcolton** [comment]
@kgriffs Do you want this to be a Talon or baked into falcon?

**Comment by sabzo** [comment]
@lwcolton When the Browser is performing the pre-flight what would be the proper way to send a "Access-Control-Allow-Origin" response? I tried the code above by @sebasmagri it set the response headers appropriately, but not for the "preflight" (https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Preflighted_requests)

**Comment by sabzo** [comment]
@sebasmagri If one is running this behind Gunicorn, does Gunicorn need to be configured for CORS? Or NGINX, if NGINX is used as a reverse proxy?

**Comment by lwcolton** [comment]
I don't believe a webserver or proxy in front of falcon needs any
configuration for CORS to work.  To respond to a preflight request, you
must implement the OPTIONS method. I am not sure if falcon provides an
on_options callback or not. If it does not, create a github issue and  I
will implement it

**Comment by lwcolton** [comment]
@kgriffs still working on this, implementation is done and I am writing tests, looking good so far

**Comment by lwcolton** [comment]
@kgriffs just wanted to say I haven't forgotten this and we will need this feature for a project we are currently working on so I am again able to allocate time to it, hoping to have a PR ready soon!

**Comment by kgriffs** [comment]
@lwcolton cool, I've also been swamped lately with some things, but hope to circle back and help test/review soon.

**Comment by advance512** [comment]
Any update? I need this, as well.

Would be nice to have it return the correct Access-Control-Allow-Methods based on the resource in question.. :)

**Comment by lwcolton** [comment]
This is still in progress, hopefully will have time to work on it in the next couple weeks.  I will also improve the handling of access-control-allow-methods to return a value based on the resource in question

**Comment by lwcolton** [comment]
The branch actually works it's just missing tests and the above mentioned feature smart access-control-allow-methods

**Comment by lwcolton** [comment]
Working on this today, thanks for your patience everyone

**Comment by lwcolton** [comment]
Alright folks we are just waiting on #607 to be fixed so this can merge.  All other tests pass.  You can view the changeset at https://github.com/HurricaneLabs/falcon/tree/feature_303 .  I will squash the changeset into a single commit message before submitting the PR.  Standby for that.

**Comment by advance512** [comment]
Sooo... what is this waiting for? :)
Do you need any help with this?

**Comment by thenano** [comment]
I'm also very keen on this feature, just wondering if there's a specific version I can use or a target release for it, and some documentation on how to use it?

Thanks!

**Comment by kgriffs** [comment]
@lwcolton Can you give us an update?

**Comment by kgriffs** [comment]
Released as a standalone add-on. See also:  https://github.com/lwcolton/falcon-cors
